src/libraries/System.Private.CoreLib/src/System/BitConverter.cs

```diff
@@ -669,25 +665,24 @@ public static string ToString(byte[] value, int startIndex, int length)
             // (int.MaxValue / 3) == 715,827,882 Bytes == 699 MB
             ArgumentOutOfRangeException.ThrowIfGreaterThan(length, int.MaxValue / 3);
 
-            return string.Create(length * 3 - 1, (value, startIndex, length), static (dst, state) =>
+            int stringLength = length * 3 - 1;
+            string result = string.FastAllocateString(stringLength);
+
+            var dst = new Span<char>(ref result.GetRawStringData(), stringLength);
+            var src = new ReadOnlySpan<byte>(value, startIndex, length);
+            int i = 0;
+            int j = 0;
+            byte b = src[i++];
+            dst[j++] = HexConverter.ToCharUpper(b >> 4);
+            dst[j++] = HexConverter.ToCharUpper(b);
+            while (i < src.Length)
             {
-                var src = new ReadOnlySpan<byte>(state.value, state.startIndex, state.length);
-
-                int i = 0;
-                int j = 0;
-
-                byte b = src[i++];
+                b = src[i++];
```

Comment:
This is one of those places where even if we changed it to:
```C#
while ((uint)i < (uint)str.Length)
```
the JIT today still wouldn't remove the bounds check for the subsequent src access, right?

Reply:
I'll see what I can do with this in net8.0 at some point, but loop-related bound checks have always been the worst (fragile) 😥

Reply:
I opened https://github.com/dotnet/runtime/issues/83349 :)